feat(terraform): update terraform lock files

secustor opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->

## Changes:

This PR adds a post update hook, similar to NPM to update Terraform lock files. 

~~A Terraform binary ( >= 0.14.0 ) has to be in PATH for this to work.~~

## Context:

~~Prerequisites: https://github.com/renovatebot/docker-renovate-full/issues/256~~ 
Closes #7895

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [x] Unit tests + ran on a real repository

https://github.com/secustor/renovate_terraform_lockfile

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/master/.github/pull_request_template.md -->

<!-- Please do not force push to this PR's branch after you have created this PR, as doing so makes it harder for us to review your work. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

viceice reviewed (changes requested):
Please do not follow npm implementation, use artifact handling like in `helmv3` manager.


https://github.com/renovatebot/renovate/blob/3ad5f98d8c44f24ea4f01a932d78de1f062508cd/lib/manager/helmv3/artifacts.ts#L23

viceice commented on lib/workers/branch/index.ts:
```diff
@@ -302,15 +302,15 @@ export async function processBranch(
       logger.debug('No package files need updating');
     }
-    const additionalFiles = await getAdditionalFiles(
+    const npmAdditionalFiles = await getAdditionalFiles(
```
This file should be reverted

viceice commented on lib/manager/terraform/post-update/index.ts:
```diff
@@ -0,0 +1,40 @@
+import { execSync } from 'child_process';
```
This file is obsolete?

secustor replied:
yeah, has been an oversight on my part

viceice reviewed (commented):
This needs some unit tests

secustor commented:
I have copied the helmv3 unit tests and adapted them to Terraform, but I`m not sure if they are correct like they are right now.

Djiit commented:
Hey guys, can we help with this in any way? Would be awesome to see this on master. Cheers

viceice commented on lib/manager/terraform/artifacts.ts:
```diff
@@ -0,0 +1,78 @@
+const lockFile = '.terraform.lock.hcl';
+
+async function terraformLockUpdate(): Promise<void> {
+  await deleteLocalFile(lockFile);
```
Why always delete? Normally we only do this for lockfile maintenance.

rarkins replied:
Do we know if terraform reuses the existing lock file, e.g. to speed up calculations or (ideally) to keep changes to a minimum where possible?

secustor replied:
This is based upon this comment https://github.com/hashicorp/terraform/issues/27161#issuecomment-742133251 in an issue which I opened to discuss such a reuse of a enxisting lockfile.     
The workaround is to delete the lockfile and then recreate one with `terraform provider lock`

viceice commented on lib/manager/terraform/artifacts.ts:
```diff
@@ -0,0 +1,78 @@
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  logger.debug(`terraform.updateArtifacts(${packageFileName})`);
+
+  const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';
+
+  if (
+    !isLockFileMaintenance &&
+    (updatedDeps === undefined || updatedDeps.length < 1)
+  ) {
+    logger.debug('No updated terraform deps - returning null');
+    return null;
+  }
+
+  const lockFileName = getSiblingFileName(packageFileName, lockFile);
+  const existingLockFileContent = await readLocalFile(lockFileName);
+  if (!existingLockFileContent) {
+    logger.debug('No .terraform.lock.hcl found');
+    return null;
+  }
```
So this get potencially run multiple times or never too.

1. never run when there is no update in lockfile folder
2. multiple runs when there are multiple updates in that folder

@rarkins how far are we with the seperate lock files properties? Maybe we can add the lockfile as seperate packageFile with only lockfile update? 🤔

rarkins replied:
I think in theory that we need to run a lock file update (a) if any deps change, or (b) if we upgrade the terraform version. Not sure though whether treating the lockfile as its own package file makes sense or fits into the above requirements

secustor replied:
If we incorporate following request https://github.com/renovatebot/renovate/issues/7895#issuecomment-754942081 we need to run it on the default branch too and if it has changed create a new branch and  PR. So basically: 
1. When a provider is updated
2. When Terraform is updated 
3. On the master/default branch to update the version in the lockFile

viceice replied:
@secustor this is what we call lockfile maintenance update. So in update artifacts you get a lockfile maintenance update type. Which should only update the lockfile to latest versions in range.

secustor commented:
Because of https://github.com/hashicorp/terraform/issues/28333 and as there is no movement regarding the Hashicorp Registry improvements,  I will investigate more in the direction to handle the lock file directly without using the terraform binary

My initial thoughts are: 
- add capabilities to datasources that multiple hashes per release can be returned as we need this to handle a hash per platform
- Download binaries, create hashes using sha256 https://pkg.go.dev/golang.org/x/mod/sumdb/dirhash && https://github.com/hashicorp/terraform/blob/main/internal/getproviders/hash.go#L349
  This can be tricky but should be possible
- heavily cache the results 
- update lock file as artifact. 

Any thoughts from you guys?

viceice commented:
Sounds good. Hashes should be safe to store in global cache for long time as they shouldn't ever change. They are also not privacy relevant.

secustor commented:
changed  the approach and dropped the hash generation at the datasource as the intial sync would take about 1-2h per provider.    
I have implemented it no at the artifact layer as we here already know which exact version is needed.

secustor commented:
@viceice Do you have access to more detailed logs?  An unit tests is failing which doesn't fail locally. My guess is that the folder, where the zip is downloaded does not exist, which should be there. 

Everything else is ready for review. 

Thx!

secustor commented:
@viceice  Thx! That seems to be the source of the issue. 

The problems seems to be that Github does not support to add LFS objects on public forks, if there is no LFS on the root configured yet. https://docs.github.com/en/github/managing-large-files/collaboration-with-git-large-file-storage#pushing-large-files-to-forks

So I think there are two possibilities: 
- somebody with write privileges add a LFS object this repo
- turn of the nock fence so that the test can directly download the file.

viceice commented:
@rarkins What do you think? adding that big file to repo (lfs) or download on the fly while testing?

secustor commented:
We are talking about 2 files, 30Mb each, but I can change the test to another provider which is smaller.     
`random` for example has 5Mb per build artifact, that would 10Mb per test run. 

This is the case if it is working that only http requests to the builds  are let through nock.    
Should it download **all** builds for the specific version we are talking about 60 Mb for the `random` provider.

rarkins commented:
5MB is tolerable, I would like to avoid too much more being added to the repo if we can.

secustor commented:
I'm not sure if this clear right now, but somebody with write access has to upload a Git LFS object to renovate before I can proceed. 

IMO the images under docs/ would be good fit for this, as it is not recommended to add binary files to GIT anyway

rarkins commented:
Each contributor of the repo will then need to install `lfs`, right? Should we have any concern that it will be a further barrier to contributors?

rarkins commented:
Here are the install guides, for reference: https://docs.github.com/en/github/managing-large-files/installing-git-large-file-storage

secustor commented:
Yeah, everybody who wants to manipulate/use the files checked in LFS would need to install it. 

Maybe I'm the outlier here, but I activate GIT LFS for every repo I create and plan to work on longterm.

rarkins commented:
My concerns are about casual/community contributors and not accidentally creating a barrier to contribution. I think people often drop off silently if they can't get a repo to work easily

secustor commented:
I have added a 500 byte zip as normal GIT object instead of an LFS object. This should be fine as long it is not changed often.

secustor commented:
@viceice Any chance that you find this week some time to take a rough look at his PR? 
I would like to now if there are still bigger changes necessary.

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -0,0 +1,188 @@
+    'terraform-provider-release',
+    cacheKey,
+    sortedHashes,
+    10080
```
can we create a module const for this until it will be configurable?

viceice commented on lib/manager/terraform/lockfile/util.spec.ts:
```diff
@@ -0,0 +1,22 @@
+import { getName } from '../../../../test/util';
+import { extractLocks } from './util';
+
+const validLockfile = readFileSync(
```
use new test util file read function

viceice commented on lib/manager/terraform/lockfile/util.ts:
```diff
@@ -0,0 +1,204 @@
+const constraintLineRegex = /^(?<prefix>[\s]*constraints[\s]*=[\s]*")(?<constraint>[^"']+)(?<suffix>".*)$/;
+const hashLineRegex = /^(?<prefix>\s*")(?<hash>[^"]+)(?<suffix>",.*)$/;
+
+const lockFile = '.terraform.lock.hcl';
```
So for now only repo root lock files are supported?

secustor replied:
Yes, but this should easy to add later on.

viceice commented on lib/manager/terraform/lockfile/util.ts:
```diff
@@ -0,0 +1,204 @@
+  ProviderSlice,
+} from './types';
+
+const providerStartLineRegex = /^provider "(?<registryUrl>[^/]*)\/(?<namespace>[^/]*)\/(?<depName>[^/]*)"/;
+const versionLineRegex = /^(?<prefix>[\s]*version[\s]*=[\s]*")(?<version>[^"']+)(?<suffix>".*)$/;
+const constraintLineRegex = /^(?<prefix>[\s]*constraints[\s]*=[\s]*")(?<constraint>[^"']+)(?<suffix>".*)$/;
+const hashLineRegex = /^(?<prefix>\s*")(?<hash>[^"]+)(?<suffix>",.*)$/;
```
@zharinov Can you create a `moo` lexer for this, so we don't need this complex regexes and parser functions?

I think  a `moo` parser would make this files a lot smaller and simplier.

zharinov replied:
Sure

zharinov replied:
Assign some issue to me maybe?

viceice replied:
There is no issue specifically, this is brand new and part of#7895

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,340 @@
+import { updateArtifacts } from './index';
+
+// auto-mock fs
+jest.mock('fs-extra');
```
I like to see the usage of our `util/fs` wrapper, so mocking should be much easier, as we can't mock `fs` .

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -0,0 +1,188 @@
+      const writeStream = createWriteStream(downloadFileName);
+      stream.pipe(writeStream);
+
+      const streamPromise = new Promise((resolve, reject) => {
```
use promisify pipeline from stream here to have full async await support. add required functions to `util/fs` when missing

```js
const stream = require('stream');
const fs = require('fs');
const zlib = require('zlib');
const util = require('util');

const pipeline = util.promisify(stream.pipeline);

async function run() {
  try {
    await pipeline(
      fs.createReadStream('The.Matrix.1080p.mkv'),
      zlib.createGzip(),
      fs.createWriteStream('The.Matrix.1080p.mkv.gz'),
    );
    console.log('Pipeline succeeded');
  } catch (err) {
    console.error('Pipeline failed', err);
  }
}
```
https://nodejs.org/en/docs/guides/backpressuring-in-streams/

viceice reviewed (commented):
See all my comments 🙃

secustor commented:
Thx for the review. I will  look into the changes in the next days.

secustor commented on lib/manager/terraform/lockfile/hash.spec.ts:
```diff
@@ -0,0 +1,77 @@
+jest.setTimeout(15000);
+
+const releaseBackendUrl = defaultRegistryUrls[1];
+const releaseBackendAzurerm = readFileSync(
```
So basically implement a wrapper for `readFileSync` in `test/util`?

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -0,0 +1,172 @@
+      );
+
+      // delete zip file
+      await fs.deleteLocalFile(downloadFileName);
+      await fs.deleteLocalFile(extractPath);
```
```suggestion
      await fs.unlink(downloadFileName);
      await fs.remove(extractPath);
```

🤔 if i remember right, the local variants are for git repo relative files only. So you need to use normal `deleteFile`.

https://github.com/renovatebot/renovate/blob/460bfe95f030fcc1f0573d1bc17bb054c5b03cc5/lib/util/fs/index.ts#L55-L60

https://github.com/renovatebot/renovate/blob/095f2f8240ca90d3dc61016cb97b84d8e853c9b8/lib/util/fs/proxies.ts#L46-L53

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -0,0 +1,172 @@
+        await pipeline(readStream, writeStream);
+      } catch (err) {
+        /* istanbul ignore next */
+        logger.error({ err }, 'write stream error');
```
Shouldn't we throw an error if this happens?

secustor replied:
I we throw an error or return null. The files and extraction directories are not getting deleted. 
And if the file is corrupted it `hashOfZip` will return null anyway

viceice replied:
🤔  OK. Maybe better use a finally block? or delete dir recusive after all files are processed?

viceice commented on lib/manager/terraform/lockfile/util.ts:
```diff
@@ -0,0 +1,204 @@
+
+export function isPinnedVersion(value: string): boolean {
+  const versioning = getVersioning('hashicorp');
+  return <boolean>versioning.isSingleVersion(value);
```
```suggestion
  return !!versioning.isSingleVersion(value);
```

do not use `<boolean>` cast use double negation or  `as boolean` cast.

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,328 @@
+      'h1:6zB2hX7YIOW26OrKsLJn0uLMnjqbPNxcz9RhlWEuuSY=',
+    ]);
+
+    httpMock
```
Always add an expect for the httpMock trace, so we can see any http layer changes which may cause errors.

secustor replied:
The problem is that the order is not deterministic. Therefor some times it will match the snapshot sometimes not.

viceice replied:
Then we should at least match the count, or order by url before making snapshot

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,328 @@
+    expect(mockHash.mock.calls).toBeArrayOfSize(1);
+    expect(mockHash.mock.calls).toMatchSnapshot();
+  });
+  it('update single dependency with range constraint and minor update', async () => {
```
```suggestion

  it('update single dependency with range constraint and minor update', async () => {
```

add newline for readabillity between tests.

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,328 @@
+import { join } from 'upath';
+import * as httpMock from '../../../../test/http-mock';
+import { getName, loadFixture } from '../../../../test/util';
```
```suggestion
import { fs, getName, loadFixture } from '../../../../test/util';
```

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,328 @@
+import { getName, loadFixture } from '../../../../test/util';
+import { getPkgReleases } from '../../../datasource';
+import { defaultRegistryUrls } from '../../../datasource/terraform-provider';
+import * as _fs from '../../../util/fs';
```
```suggestion
```

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,328 @@
+  '../../../datasource/terraform-provider'
+);
+
+const fs: jest.Mocked<typeof _fs> = _fs as any;
```
```suggestion
```

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,328 @@
+jest.mock('./hash');
+jest.mock('../../../datasource');
+
+jest.setTimeout(15000);
```
Are the tests that slow?

secustor replied:
no, this a left over from when I have used bigger files and haven't mocked the FS operations. removing

viceice reviewed (commented):
next round 😅

viceice commented on lib/manager/terraform/lockfile/util.ts:
```diff
@@ -0,0 +1,204 @@
+      contents: newContent,
+    },
+  };
+  return new Promise<UpdateArtifactsResult[]>((resolve) => resolve([result]));
```
```suggestion
  return Promise.resolve([result]);
```
Simplify. Better we make this function sync.

secustor replied:
Is it really 'better' or only simpler. It's a genuine question.

viceice replied:
We prefer simpler code, because it's easier to read and understand. 😏

secustor replied:
K, I thought maybe there are other implications to this in sync here

viceice commented on lib/util/fs/index.ts:
```diff
@@ -133,3 +135,21 @@ export async function findLocalSiblingOrParent(
+ * Get files by name from directory
+ */
+export function readLocalDirectory(path: string): Promise<string[]> {
+  return fs.readdir(path);
```
```suggestion
  const localPath = join(localDir, path);
  return fs.readdir(localPath);
```

Local function only work on repo relative files. Normal fs function should added to `./proxies.ts` and can be path through from `fs-extra`

viceice commented on lib/util/fs/index.ts:
```diff
@@ -133,3 +135,23 @@ export async function findLocalSiblingOrParent(
+  return fs.createWriteStream(path);
+}
+
+export function getStreamingPipeline(): (
+  arg1: NodeJS.ReadableStream,
+  arg2: NodeJS.ReadWriteStream | NodeJS.WritableStream
+) => Promise<void> {
+  return util.promisify(stream.pipeline);
+}
```
```suggestion
export const pipeline = util.promisify(stream.pipeline);
```

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -0,0 +1,170 @@
+      );
+      const readStream = http.stream(build.url);
+      const writeStream = fs.createWriteStream(downloadFileName);
+      readStream.pipe(writeStream);
+
+      const pipeline = fs.getStreamingPipeline();
```
```suggestion
```

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -0,0 +1,170 @@
+
+      let hash = null;
+      try {
+        await pipeline(readStream, writeStream);
```
```suggestion
        await fs.pipeline(readStream, writeStream);
```

This will do what we want. checkout https://nodejs.org/docs/latest-v14.x/api/stream.html#stream_stream_pipeline_streams_callback

See also my changes to `util/fs`

viceice commented on lib/manager/terraform/lockfile/hash.ts:
```diff
@@ -0,0 +1,170 @@
+      } finally {
+        // delete zip file
+        await fs.unlink(downloadFileName);
+        await fs.remove(extractPath);
```
Isn't this already removed by `hashOfZipContent` ? so we can do it only once here?
```suggestion
        await fs.rm(extractPath, { recursive: true });
```

viceice commented on lib/manager/terraform/lockfile/index.ts:
```diff
@@ -0,0 +1,107 @@
+  );
+
+  const result = updates.filter((value) => value);
+  return new Promise((resolve) => resolve(result));
```
```suggestion
  return result;
```
simplify.

viceice reviewed (commented):
Some more improvements 🙃

viceice commented on lib/manager/terraform/lockfile/index.ts:
```diff
@@ -0,0 +1,106 @@
+        datasource: 'terraform-provider',
+        depName: lock.lookupName,
+      };
+      const releasesResult = await getPkgReleases(updateConfig);
+      const releases = releasesResult.releases;
```
```suggestion
      const { releases } = await getPkgReleases(updateConfig);
```

just a small simplification

viceice commented on lib/manager/terraform/lockfile/index.ts:
```diff
@@ -0,0 +1,106 @@
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  logger.debug(`terraform.updateArtifacts(${packageFileName})`);
+
+  const lockFileContent = await readLockFile();
```
We can use `fs.findLocalSiblingOrParent` later to search upwards from `packageFileName`, but maybe `fs.getSiblingFileName` would be enough as the lock file should be in same folder.

https://github.com/renovatebot/renovate/blob/460bfe95f030fcc1f0573d1bc17bb054c5b03cc5/lib/util/fs/index.ts#L108-L117

secustor replied:
I'm if this would work in mono repos, but I think we should ship it like this and gather feedback first.

secustor replied:
If it's you ^^ ...

viceice commented on lib/manager/terraform/lockfile/util.ts:
```diff
@@ -0,0 +1,206 @@
+      contents: newContent,
+    },
+  };
+  return Promise.resolve([result]);
```
Not happy about that but ok.

secustor replied:
Same, but not sure how to solve this otherwise

viceice replied:
just make this function not async? As it has no awaits or other promis handling.

secustor replied:
The function one up in the stack requires a pomised result or null

viceice replied:
```suggestion
  return result;
```

viceice reviewed (commented):
jusr some small things left

viceice commented on lib/manager/terraform/lockfile/util.ts:
```diff
@@ -0,0 +1,206 @@
+export function writeLockUpdates(
+  updates: ProviderLockUpdate[],
+  oldLockFileContent: string
+): Promise<UpdateArtifactsResult[]> {
```
```suggestion
): UpdateArtifactsResult {
```

viceice commented on lib/manager/terraform/lockfile/index.ts:
```diff
@@ -0,0 +1,106 @@
+    return null;
+  }
+
+  return writeLockUpdates(updates, lockFileContent);
```
```suggestion
  return [writeLockUpdates(updates, lockFileContent)];
```
```suggestion
  const res = writeLockUpdates(updates, lockFileContent);
  return res ? [ res ] : null;
```

viceice reviewed (commented):
This should work now

viceice commented on lib/manager/terraform/lockfile/index.ts:
```diff
@@ -0,0 +1,106 @@
+export async function updateArtifacts({
+  packageFileName,
+  updatedDeps,
+  newPackageFileContent,
```
```suggestion
```

not used, so remove

viceice commented on lib/manager/terraform/lockfile/util.ts:
```diff
@@ -0,0 +1,207 @@
+import * as fs from '../../../util/fs';
+import { readLocalFile } from '../../../util/fs';
```
Please use only one import.

viceice reviewed (commented):
Needs deconflicting

rarkins commented:
@secustor @viceice what's the next step here?

rarkins commented:
Is anything still needed in containerbase for this to work with binarySource=docker?

viceice commented:
Yes, we need a terraform docker image. 🤦‍♂️

rarkins commented:
I can't find this unresolved conversation so may have to use admin privileges to override this check:

![image](https://user-images.githubusercontent.com/6311784/120463047-ceb13800-c39b-11eb-921c-f548dad1be20.png)

secustor commented:
No, there is no need anymore for a Terraform binary or at least not for this PR.     

The update logic is completely implemented inside of Renovate.

viceice reviewed (approved):
Right, little confused, too much sun yesterday. 😅

viceice commented:
@rarkins maybe hide behind an experimental env flag?

rarkins commented:
> @rarkins maybe hide behind an experimental env flag?

Because we're worried it may not fully work, or because some people may not want it?

viceice commented:
I think it can produce a lot of load to renovate and the terraform registry (downloads).

So i would suggest to test it out in wild slowly and general enable it in a future pr?

secustor commented:
It will download the provider binaries for all platforms of the requested provider version at least once which will at least initially impact the traffic size.   

But I think we should put it behind a feature flag initially just in case there are edge cases which are not documented on the Terraform side of things.

rarkins reviewed (changes requested):
Needs configuration option added, set to false

viceice commented on docs/usage/self-hosted-experimental.md:
```diff
@@ -35,6 +35,10 @@ If set to any value, Renovate will always paginate requests to GitHub fully, ins
 
 If set to "false" (string), Renovate will remove any existing `package-lock.json` before attempting to update it.
 
+## RENOVATE_TERRAFORM_LOCK_FILE
```
```suggestion
## RENOVATE_X_TERRAFORM_LOCK_FILE
```
Please use `RENOVATE_X_` prefix for experimental options, as that is the new standard prefix.

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,411 @@
+
+  afterEach(() => {
+    setAdminConfig();
+    delete process.env.RENOVATE_TERRAFORM_LOCK_FILE;
```
```suggestion
    delete process.env.RENOVATE_X_TERRAFORM_LOCK_FILE;
```

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,411 @@
+  it('returns null if no .terraform.lock.hcl found', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(null);
+
+    process.env.RENOVATE_TERRAFORM_LOCK_FILE = 'test';
```
```suggestion
    process.env.RENOVATE_X_TERRAFORM_LOCK_FILE = 'test';
```

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,411 @@
+  it('returns null if .terraform.lock.hcl is empty', async () => {
+    fs.readLocalFile.mockResolvedValueOnce('empty' as any);
+
+    process.env.RENOVATE_TERRAFORM_LOCK_FILE = 'test';
```
```suggestion
    process.env.RENOVATE_X_TERRAFORM_LOCK_FILE = 'test';
```

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,411 @@
+      ...config,
+    };
+
+    process.env.RENOVATE_TERRAFORM_LOCK_FILE = 'test';
```
```suggestion
    process.env.RENOVATE_X_TERRAFORM_LOCK_FILE = 'test';
```

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,411 @@
+      ...config,
+    };
+
+    process.env.RENOVATE_TERRAFORM_LOCK_FILE = 'test';
```
```suggestion
    process.env.RENOVATE_X_TERRAFORM_LOCK_FILE = 'test';
```

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,411 @@
+      ...config,
+    };
+
+    process.env.RENOVATE_TERRAFORM_LOCK_FILE = 'test';
```
```suggestion
    process.env.RENOVATE_X_TERRAFORM_LOCK_FILE = 'test';
```

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,411 @@
+        )
+      )
+      .mockResolvedValueOnce(
+        new Promise((resolve) =>
```
Why using a promise here? 🤔 `mockResolved` will Auto relate a promise.

secustor replied:
If I am not setting it up like that, the order of the results will become randomized and therefore the `matchSnapshot` function will fail from time to time.

viceice replied:
Nope, jest makes sure `mockResolvedValueOnce` calls will be used in reverse order. So last mock will be used first. 😉 

See my commit, it works for me.

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,411 @@
+      ...config,
+    };
+
+    process.env.RENOVATE_TERRAFORM_LOCK_FILE = 'test';
```
```suggestion
    process.env.RENOVATE_X_TERRAFORM_LOCK_FILE = 'test';
```

viceice commented on lib/manager/terraform/lockfile/index.spec.ts:
```diff
@@ -0,0 +1,411 @@
+      ...config,
+    };
+
+    process.env.RENOVATE_TERRAFORM_LOCK_FILE = 'test';
```
```suggestion
    process.env.RENOVATE_X_TERRAFORM_LOCK_FILE = 'test';
```
Wouldn't be easier to set in beforeEach once for all tests?

viceice commented on lib/manager/terraform/lockfile/index.ts:
```diff
@@ -0,0 +1,122 @@
+  logger.debug(`terraform.updateArtifacts(${packageFileName})`);
+
+  // TODO remove experimental flag, if functionality is confirmed
+  if (!process.env.RENOVATE_TERRAFORM_LOCK_FILE) {
```
```suggestion
  if (!process.env.RENOVATE_X_TERRAFORM_LOCK_FILE) {
```

secustor commented:
@viceice Any chance you find some time this week to take a look?

JamieMagee commented:
Merged latest `main` into this branch, just to be sure that there were no conflicts with #10362.

secustor commented:
There have been conflicts, but they have been resolved in 051506c

JamieMagee commented:
@secustor Ah, I'm sorry! When I initially looked I was only concerned with git conflicts, and didn't look for implementation conflicts. I hope it wasn't too much difficulty.

rarkins commented:
So the first use case for this will be self-hosted only, right? Because it's a env variable then we app users can't opt-in individually. Could be good to have @secustor battle test it a little first anyway :)

secustor commented:
@JamieMagee No problem! I have found this out by accident as I have been merging the main. 

@rarkins  Yes that would work only on self hosted instances. Regarding battle testing, I'm not using lock files as we pin the deps to exact versions. But I'm sure there will be people using it as it has been the most upvoted issue.

rarkins commented:
I think the tests may fail on Windows:

![image](https://user-images.githubusercontent.com/6311784/122193701-01c3f300-ce95-11eb-8751-94ad32b74ff8.png)

(we only run Windows tests on `main`, not every PR)

We'll need to fix ASAP or revert.
